spec/02-integration/05-proxy/01-router_spec.lua

```diff
@@ -387,4 +387,206 @@ describe("Router", function()
+        local i = 1
+
+        if strip_uri == true then
+          i = i + 3
+        elseif strip_uri == false then
+          i = i + 6
+        end
+
+        if upstream_uri ~= "/" then
+          i = i + 1
+          if string.sub(upstream_uri, -1) == "/" then
+            i = i + 1
+          end
+        end
+
+        local res = assert(client:send {
+          method  = "GET",
+          path    = request_uri,
+          headers = {
+            ["Host"] = "localbin-" .. i .. ".com",
+          }
+        })
+
+        local json = assert.res_status(200, res)
+        local data = cjson.decode(json)
+
+        assert.equal(expected_uri, data.vars.request_uri)
+      end
+    end
+
+    local checks = {
+      { "/",         "/",         "/",                 nil   },
+      { "/",         "/foo/bar",  "/",                 nil   },
+      { "/",         "/foo/bar/", "/",                 nil   },
+      { "/foo/bar",  "/",         "/foo/bar",          nil   },
+      { "/foo/bar/", "/",         "/foo/bar/",         nil   },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar",          nil   },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar",          nil   },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/",         nil   },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/",         nil   },
+      { "/",         "/",         "/",                 true  },
+      { "/",         "/foo/bar",  "/",                 true  },
+      { "/",         "/foo/bar/", "/",                 true  },
+      { "/foo/bar",  "/",         "/foo/bar",          true  },
+      { "/foo/bar/", "/",         "/foo/bar/",         true  },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar",          true  },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar",          true  },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/",         true  },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/",         true  },
+      { "/",         "/",         "/",                 false },
+      { "/",         "/foo/bar",  "/foo/bar",          false },
+      { "/",         "/foo/bar/", "/foo/bar/",         false },
+      { "/foo/bar",  "/",         "/foo/bar",          false },
+      { "/foo/bar/", "/",         "/foo/bar/",         false },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar/foo/bar",  false },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar/foo/bar",  false },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/foo/bar/", false },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/foo/bar/", false },
+    }
+
+    local unpack = table.unpack or unpack
+
+    for _, args in ipairs(checks) do
+
+      local config = "(strip_uri = n/a)"
+
+      if args[4] == true then
+        config = "(strip_uri = on) "
+      elseif args[4] == false then
+        config = "(strip_uri = off)"
+      end
+
+      local space = string.sub(args[1], -1) == "/" and "" or " "
+
+      it(config .. " is not appended to upstream uri " .. args[1] .. space .. " when requesting " .. args[2],
+        check(unpack(args)))
+    end
+  end)
```

Comment:
Just a quick note, router specs are not going through the whole pipeline to the API/service, hence it really does not test that slash at the end really behaves correctly all the way to the end. Router is just a one component in this equation. There are Nginx configs, proxy etc. in addition.